From 247a8266d24a01ded003f7bdbcfee86b2174bf45 Mon Sep 17 00:00:00 2001
From: Jiankang Chen <chenjiankang1@huawei.com>
Date: Tue, 29 Nov 2022 13:57:34 +0000
Subject: [PATCH] iort: Read ACPI configure to get streamid.
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: 247a8266d24a01ded003f7bdbcfee86b2174bf45
Modified-by-SEL: No


hulk inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I63SDZ

-------------------

In ascend feature, the SMMU's sid will read from acpi tabels's streamid. To
enable the SMMU, we must get SID from ACPI table, and make stall_enabled be
true.

Signed-off-by: Jiankang Chen <chenjiankang1@huawei.com>
Signed-off-by: Fang Lijun <fanglijun3@huawei.com>
Reviewed-by: Hanjun Guo <guohanjun@huawei.com>
Reviewed-by: Zhen Lei <thunder.leizhen@huawei.com>
Signed-off-by: Yang Yingliang <yangyingliang@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
index 9e702788e90f..74886facbcab 100644
--- a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
+++ b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
@@ -3249,6 +3249,9 @@ static struct iommu_device *arm_smmu_probe_device(struct device *dev)
 	struct arm_smmu_device *smmu;
 	struct arm_smmu_master *master;
 	struct iommu_fwspec *fwspec = dev_iommu_fwspec_get(dev);
+#ifdef CONFIG_ASCEND_FEATURES
+	u32 sid;
+#endif
 
 	if (!fwspec || fwspec->ops != &arm_smmu_ops)
 		return ERR_PTR(-ENODEV);
@@ -3295,6 +3298,15 @@ static struct iommu_device *arm_smmu_probe_device(struct device *dev)
 	    smmu->features & ARM_SMMU_FEAT_STALL_FORCE)
 		master->stall_enabled = true;
 
+#ifdef CONFIG_ASCEND_FEATURES
+	if (!acpi_dev_prop_read_single(ACPI_COMPANION(dev),
+			"streamid", DEV_PROP_U32, &sid)) {
+		if (iommu_fwspec_add_ids(dev, &sid, 1))
+			dev_info(dev, "failed to add ids\n");
+		master->stall_enabled = true;
+		master->ssid_bits = 0x10;
+	}
+#endif
 	arm_smmu_init_pri(master);
 
 	return &smmu->iommu;
-- 
2.33.0

